File: app/src/main/java/com/jakebarnby/filemanager/sources/onedrive/OneDriveFactory.java
package com.jakebarnby.filemanager.sources.onedrive;
import android.content.Context;

import com.jakebarnby.filemanager.sources.models.SourceStorageStats;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.PreferenceUtils;
import com.jakebarnby.filemanager.util.Utils;
import com.microsoft.graph.core.ClientException;
import com.microsoft.graph.extensions.DriveItem;
import com.microsoft.graph.extensions.Folder;
import com.microsoft.graph.extensions.IGraphServiceClient;
import com.microsoft.graph.extensions.Quota;
import com.microsoft.graph.http.GraphServiceException;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * Created by Jake on 6/9/2017.
 */

public class OneDriveFactory {
    private static final OneDriveFactory ourInstance = new OneDriveFactory();

    public static OneDriveFactory getInstance() {
        return ourInstance;
    }

    private IGraphServiceClient mGraphClient;

    private OneDriveFactory() {
    }

    public IGraphServiceClient getGraphClient() {
        return mGraphClient;
    }

    public void setGraphClient(IGraphServiceClient mGraphClient) {
        this.mGraphClient = mGraphClient;
    }

    /**
     *
     * @param id
     * @param filename
     * @param destinationPath
     * @return
     */
    public File downloadFile(String id, String filename, String destinationPath) throws IOException {
        File file = new File(destinationPath, filename);

        try(InputStream inputStream = mGraphClient
                                    .getMe()
                                    .getDrive()
                                    .getItems(id)
                                    .getContent()
                                    .buildRequest()
                                    .get()) {
            Utils.copyInputStreamToFile(inputStream, file);
            return file;
        } catch (IOException | ClientException e) {
            throw e;
        }
    }

    /**
     *  @param filePath
     * @param fileName
     * @param parentId
     */
    public DriveItem uploadFile(String filePath, String fileName, String parentId) throws IOException, GraphServiceException {
        File file = new File(filePath);

        try(FileInputStream in = new FileInputStream(file)) {
            byte[] buffer = new byte[(int)file.length()];
            in.read(buffer);

            return mGraphClient
                    .getMe()
                    .getDrive()
                    .getItems(parentId)
                    .getChildren(transformFileName(fileName))
                    .getContent()
                    .buildRequest()
                    .put(buffer);
        } catch (IOException e) {
            throw e;
        }
    }

    /**
     *
     * @param driveId
     */
    public void deleteFile(String driveId) throws GraphServiceException {
        mGraphClient
                .getMe()
                .getDrive()
                .getItems(driveId)
                .buildRequest()
                .delete();
    }

    /**
     *  @param name
     * @param parentId
     */
    public DriveItem createFolder(String name, String parentId) throws GraphServiceException {
        DriveItem item = new DriveItem();
        item.name = name;
        item.folder = new Folder();

        return mGraphClient
                .getMe()
                .getDrive()
                .getItems(parentId)
                .getChildren()
                .buildRequest()
                .post(item);
    }

    public DriveItem rename(String newName, String itemId) throws GraphServiceException {
        DriveItem item = new DriveItem();
        item.name = newName;
        return mGraphClient
                .getMe()
                .getDrive()
                .getItems(itemId)
                .buildRequest()
                .patch(item);
    }

    public SourceStorageStats getStorageStats() {
        try {
            Quota quota = mGraphClient
                    .getMe()
                    .getDrive()
                    .buildRequest()
                    .select("quota")
                    .get()
                    .quota;

            SourceStorageStats info = new SourceStorageStats();
            info.setTotalSpace(quota.total);
            info.setUsedSpace(quota.used);
            info.setFreeSpace(quota.remaining);

            return info;
        } catch (GraphServiceException e) {
            e.printStackTrace();
        }
        return null;
    }

    private String transformFileName(String fileName) {
        String newName = fileName.replaceAll(Constants.Sources.ONEDRIVE_INVALID_CHARS, "%20");
        if (newName.length() > Constants.Sources.MAX_FILENAME_LENGTH) {
            newName = newName.substring(0, Constants.Sources.MAX_FILENAME_LENGTH);
        }
        return newName;
    }

    public void logout(Context context) {
        PreferenceUtils.savePref(context, Constants.Prefs.ONEDRIVE_TOKEN_KEY, (String) null);
        PreferenceUtils.savePref(context, Constants.Prefs.ONEDRIVE_NAME_KEY, (String) null);
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/sources/onedrive/OneDriveFactory.kt
package com.jakebarnby.filemanager.sources.onedrive

import android.content.Context
import com.jakebarnby.filemanager.sources.models.SourceStorageStats
import com.jakebarnby.filemanager.util.Constants.Prefs
import com.jakebarnby.filemanager.util.Constants.Sources
import com.jakebarnby.filemanager.util.PreferenceUtils
import com.jakebarnby.filemanager.util.Utils
import com.microsoft.graph.core.ClientException
import com.microsoft.graph.extensions.DriveItem
import com.microsoft.graph.extensions.Folder
import com.microsoft.graph.extensions.IGraphServiceClient
import com.microsoft.graph.http.GraphServiceException
import java.io.File
import java.io.FileInputStream
import java.io.IOException

/**
 * Created by Jake on 6/9/2017.
 */
object OneDriveFactory {

    var service: IGraphServiceClient? = null

    /**
     *
     * @param id
     * @param filename
     * @param destinationPath
     * @return
     */
    @Throws(IOException::class)
    fun downloadFile(
        id: String,
        filename: String,
        destinationPath: String
    ): File? {
        val file = File(destinationPath, filename)
        try {
            service
                ?.me
                ?.drive
                ?.getItems(id)
                ?.content
                ?.buildRequest()
                ?.get()
                ?.use {
                    Utils.copyInputStreamToFile(it, file)
                    return file
                }
        } catch (e: IOException) {
            throw e
        } catch (e: ClientException) {
            throw e
        }
        return null
    }

    /**
     * @param filePath
     * @param fileName
     * @param parentId
     */
    @Throws(IOException::class, GraphServiceException::class)
    fun uploadFile(
        filePath: String,
        fileName: String,
        parentId: String
    ): DriveItem? {
        val file = File(filePath)
        try {
            FileInputStream(file).use {
                val buffer = ByteArray(file.length().toInt())
                it.read(buffer)
                return service
                    ?.me
                    ?.drive
                    ?.getItems(parentId)
                    ?.getChildren(transformFileName(fileName))
                    ?.content
                    ?.buildRequest()
                    ?.put(buffer)
            }
        } catch (e: IOException) {
            throw e
        }
    }

    /**
     *
     * @param driveId
     */
    @Throws(GraphServiceException::class)
    fun deleteFile(driveId: String?) {
        service
            ?.me
            ?.drive
            ?.getItems(driveId)
            ?.buildRequest()
            ?.delete()
    }

    /**
     * @param name
     * @param parentId
     */
    @Throws(GraphServiceException::class)
    fun createFolder(
        name: String,
        parentId: String
    ): DriveItem? {
        val item = DriveItem().apply {
            this.name = name
            this.folder = Folder()
        }

        return service
            ?.me
            ?.drive
            ?.getItems(parentId)
            ?.children
            ?.buildRequest()
            ?.post(item)
    }

    @Throws(GraphServiceException::class)
    fun rename(
        newName: String,
        itemId: String
    ): DriveItem? {
        val item = DriveItem().apply {
            name = newName
        }
        return service
            ?.getMe()
            ?.drive
            ?.getItems(itemId)
            ?.buildRequest()
            ?.patch(item)
    }

    val storageStats: SourceStorageStats?
        get() {
            try {
                val quota = service
                    ?.me
                    ?.drive
                    ?.buildRequest()
                    ?.select("quota")
                    ?.get()
                    ?.quota

                return SourceStorageStats().apply {
                    totalSpace += quota?.total ?: 0
                    usedSpace += quota?.used ?: 0
                    freeSpace += quota?.remaining ?: 0
                }
            } catch (e: GraphServiceException) {
                e.printStackTrace()
            }
            return null
        }

    private fun transformFileName(fileName: String): String {
        var newName = fileName.replace(Sources.ONEDRIVE_INVALID_CHARS.toRegex(), "%20")
        if (newName.length > Sources.MAX_FILENAME_LENGTH) {
            newName = newName.substring(0, Sources.MAX_FILENAME_LENGTH)
        }
        return newName
    }

    fun logout(context: Context) {
        PreferenceUtils.savePref(context, Prefs.ONEDRIVE_TOKEN_KEY, null as String?)
        PreferenceUtils.savePref(context, Prefs.ONEDRIVE_NAME_KEY, null as String?)
    }
}